File: web/js/tilemap.js
// tilemap.js — Roderick Tron | MagmaCrunch Media © 2026
// Level parsing, and the collision the whole game stands on.

const TILE_AIR = '.';
const TILE_SOLID = '#';
const TILE_PLATFORM = '=';
const TILE_WATER = '~';
const TILE_UPDRAFT = '^';
const TILE_RAIL = '-';

function isSolidTile(ch) { return ch === TILE_SOLID; }
function isPlatformTile(ch) { return ch === TILE_PLATFORM || ch === TILE_RAIL; }
function isRailTile(ch) { return ch === TILE_RAIL; }
function isWaterTile(ch) { return ch === TILE_WATER; }
function isUpdraftTile(ch) { return ch === TILE_UPDRAFT; }

/**
 * A parsed level: a grid of terrain, plus the things that were sitting in it.
 *
 * Entity characters are lifted out at parse time and their cells become air, so
 * collision never has to know what a gargoyle is.
 */
function Tilemap(def) {
    this.name = def.name;
    this.subtitle = def.subtitle || '';

    const width = Math.max.apply(null, def.rows.map(function (r) { return r.length; }));
    this.cols = width;
    this.rows = def.rows.length;
    this.w = this.cols * CONFIG.TILE;
    this.h = this.rows * CONFIG.TILE;

    this.spawn = { x: CONFIG.TILE, y: CONFIG.TILE };
    this.exit = null;
    this.notes = [];
    this.letters = [];
    this.enemies = [];
    this.bells = [];
    this.trolleys = [];

    this.grid = [];
    for (let ty = 0; ty < this.rows; ty++) {
        // Short rows are padded rather than rejected: trimming trailing air
        // from a line should not shift the rest of the level sideways.
        const row = def.rows[ty].padEnd(width, TILE_AIR).split('');
        for (let tx = 0; tx < width; tx++) {
            const ch = row[tx];
            const px = tx * CONFIG.TILE;
            const py = ty * CONFIG.TILE;
            let consumed = true;

            if (ch === 'S') {
                // Bottom-aligned in its tile, so a spawn marker drawn sitting
                // on the ground puts his feet on that ground.
                this.spawn = { x: px + 1, y: py + CONFIG.TILE - CONFIG.PLAYER_H };
            } else if (ch === 'E') {
                this.exit = { x: px, y: py - CONFIG.TILE, w: CONFIG.TILE, h: CONFIG.TILE * 2 };
            } else if (ch === 'o') {
                this.notes.push({ x: px + 4, y: py + 4, w: 8, h: 8, taken: false });
            } else if (ch === 'T' || ch === 'R' || ch === 'O' || ch === 'N') {
                this.letters.push({
                    ch: ch, x: px + 2, y: py + 2,
                    w: CONFIG.TILE - 4, h: CONFIG.TILE - 4, taken: false,
                });
            } else if (ch === 'g') {
                this.enemies.push({ kind: 'gargoyle', x: px, y: py + CONFIG.TILE - CONFIG.GARGOYLE_H });
            } else if (ch === 'f') {
                this.enemies.push({ kind: 'flyer', x: px, y: py });
            } else if (ch === 'b') {
                this.bells.push({
                    x: px + (CONFIG.TILE - CONFIG.BELL_W) / 2,
                    y: py + (CONFIG.TILE - CONFIG.BELL_H) / 2,
                });
            } else if (ch === 'c') {
                this.trolleys.push({ x: px, y: py + CONFIG.TILE - CONFIG.TROLLEY_H });
            } else if (ch === 's') {
                this.enemies.push({ kind: 'statue', x: px, y: py + CONFIG.TILE - CONFIG.GARGOYLE_H });
            } else {
                consumed = false;
            }

            if (consumed) row[tx] = TILE_AIR;
        }
        this.grid.push(row);
    }
}

/**
 * The character at a tile.
 *
 * Off the left and right edges reads as solid, so a level is walled rather than
 * open to an endless sideways fall. Above and below read as air: dropping out
 * of the bottom is how a pit kills you.
 */
Tilemap.prototype.tileAt = function (tx, ty) {
    if (tx < 0 || tx >= this.cols) return TILE_SOLID;
    if (ty < 0 || ty >= this.rows) return TILE_AIR;
    return this.grid[ty][tx];
};

/** Every tile a pixel-space box touches, as [tx0, ty0, tx1, ty1]. */
Tilemap.prototype.tileRange = function (x, y, w, h) {
    return [
        Math.floor(x / CONFIG.TILE),
        Math.floor(y / CONFIG.TILE),
        Math.floor((x + w - 1) / CONFIG.TILE),
        Math.floor((y + h - 1) / CONFIG.TILE),
    ];
};

Tilemap.prototype.overlaps = function (x, y, w, h, test) {
    const r = this.tileRange(x, y, w, h);
    for (let ty = r[1]; ty <= r[3]; ty++) {
        for (let tx = r[0]; tx <= r[2]; tx++) {
            if (test(this.tileAt(tx, ty))) return true;
        }
    }
    return false;
};

/** Does this box overlap brick? One-way platforms are not solid here. */
Tilemap.prototype.overlapsSolid = function (x, y, w, h) {
    return this.overlaps(x, y, w, h, isSolidTile);
};

Tilemap.prototype.overlapsWater = function (x, y, w, h) {
    return this.overlaps(x, y, w, h, isWaterTile);
};

/** Is any part of this box in rising air? Updraft tiles are not solid. */
Tilemap.prototype.overlapsUpdraft = function (x, y, w, h) {
    return this.overlaps(x, y, w, h, isUpdraftTile);
};

/** Horizontal centre of the updraft column this box is in, or null. */
Tilemap.prototype.updraftCentre = function (x, y, w, h) {
    const r = this.tileRange(x, y, w, h);
    for (let ty = r[1]; ty <= r[3]; ty++) {
        for (let tx = r[0]; tx <= r[2]; tx++) {
            if (isUpdraftTile(this.tileAt(tx, ty))) return tx * CONFIG.TILE + CONFIG.TILE / 2;
        }
    }
    return null;
};

/**
 * Top of the highest one-way platform this box has just landed on, or null.
 *
 * `prevBottom` is the whole of what makes it one-way: a platform only stops a
 * body that was already above it, so you rise through the same tile you later
 * stand on.
 */
Tilemap.prototype.platformTop = function (x, y, w, h, prevBottom) {
    const r = this.tileRange(x, y, w, h);
    // Scan one row further down than the body occupies. tileRange stops at
    // (y + h - 1), so a body resting EXACTLY on a platform — bottom flush with
    // the tile top — never sees the tile holding it up, and sinks: the first
    // frame the tile is out of range, and by the second the playhead has passed
    // it and the came-from-above guard rejects it. Landing from a height hides
    // this, because a fast step crosses the boundary in one move.
    const bottomRow = Math.floor((y + h) / CONFIG.TILE);
    let best = null;
    for (let ty = r[1]; ty <= Math.max(r[3], bottomRow); ty++) {
        const top = ty * CONFIG.TILE;
        if (prevBottom > top + 0.001) continue;        // came from below, or inside
        for (let tx = r[0]; tx <= r[2]; tx++) {
            if (isPlatformTile(this.tileAt(tx, ty)) && (best === null || top < best)) best = top;
        }
    }
    return best;
};

/**
 * Move a body horizontally, stopping flush against brick.
 *
 * Stepped under a tile at a time so nothing can tunnel through a wall. Normal
 * running never needs it — 4.2px a frame at most — but a stomp bounce plus
 * knockback can exceed a tile in one step, and passing through a wall is the
 * one bug that makes a platformer feel broken rather than merely hard.
 */
Tilemap.prototype.moveX = function (box, dx) {
    if (!dx) return false;
    const cap = CONFIG.TILE - 1;
    const step = Math.sign(dx) * Math.min(Math.abs(dx), cap);
    let remaining = dx;
    let hit = false;

    while (Math.abs(remaining) > 0.0001) {
        const move = Math.abs(remaining) < Math.abs(step) ? remaining : step;
        const next = box.x + move;
        if (this.overlapsSolid(next, box.y, box.w, box.h)) {
            box.x = move > 0
                ? Math.floor((next + box.w) / CONFIG.TILE) * CONFIG.TILE - box.w
                : (Math.floor(next / CONFIG.TILE) + 1) * CONFIG.TILE;
            hit = true;
            break;
        }
        box.x = next;
        remaining -= move;
    }
    return hit;
};

/**
 * Move a body vertically. Reports which surface stopped it.
 *
 * `dropping` suppresses one-way platforms, for a body deliberately dropping
 * through one.
 */
Tilemap.prototype.moveY = function (box, dy, dropping) {
    const result = { ground: false, ceiling: false, platform: false };
    if (!dy) return result;

    const cap = CONFIG.TILE - 1;
    const step = Math.sign(dy) * Math.min(Math.abs(dy), cap);
    let remaining = dy;

    while (Math.abs(remaining) > 0.0001) {
        const move = Math.abs(remaining) < Math.abs(step) ? remaining : step;
        const prevBottom = box.y + box.h;
        const next = box.y + move;

        if (this.overlapsSolid(box.x, next, box.w, box.h)) {
            if (move > 0) {
                box.y = Math.floor((next + box.h) / CONFIG.TILE) * CONFIG.TILE - box.h;
                result.ground = true;
            } else {
                box.y = (Math.floor(next / CONFIG.TILE) + 1) * CONFIG.TILE;
                result.ceiling = true;
            }
            break;
        }

        if (move > 0 && !dropping) {
            const top = this.platformTop(box.x, next, box.w, box.h, prevBottom);
            if (top !== null) {
                box.y = top - box.h;
                result.ground = true;
                result.platform = true;
                break;
            }
        }

        box.y = next;
        remaining -= move;
    }
    return result;
};

/** Is this box sitting over rail? The trolley only rolls where there is track. */
Tilemap.prototype.overlapsRail = function (x, y, w, h) {
    return this.overlaps(x, y, w, h, isRailTile);
};

/** Is there footing directly under this box? Used by enemy ledge-turning. */
Tilemap.prototype.hasFooting = function (x, y, w, h) {
    return this.overlapsSolid(x, y + h, w, 1)
        || this.overlaps(x, y + h, w, 1, isPlatformTile);
};

File: tui/tron/tilemap.py
"""Level parsing, and the collision the whole game stands on.

A direct port of ``web/js/tilemap.js``. It imports nothing outside the standard
library and knows nothing about terminals: everything here is world pixels.

The one structural difference from the JavaScript is that boxes are a real
type (:class:`Box`) rather than object literals, because :meth:`Tilemap.move_x`
and :meth:`Tilemap.move_y` mutate the thing they are handed and a dataclass
says so.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from . import config as C

TILE_AIR = "."
TILE_SOLID = "#"
TILE_PLATFORM = "="
TILE_WATER = "~"
TILE_UPDRAFT = "^"
TILE_RAIL = "-"

#: Characters lifted out of the grid at parse time. Their cells become air, so
#: collision never has to know what a gargoyle is.
ENTITY_CHARS = frozenset("SEoTRONgfsbc")


def is_solid_tile(ch: str) -> bool:
    return ch == TILE_SOLID


def is_platform_tile(ch: str) -> bool:
    return ch in (TILE_PLATFORM, TILE_RAIL)


def is_rail_tile(ch: str) -> bool:
    return ch == TILE_RAIL


def is_water_tile(ch: str) -> bool:
    return ch == TILE_WATER


def is_updraft_tile(ch: str) -> bool:
    return ch == TILE_UPDRAFT


def sign(n: float) -> float:
    """JavaScript's ``Math.sign``: -1, 0 or 1, and 0 stays 0."""
    if n > 0:
        return 1.0
    if n < 0:
        return -1.0
    return 0.0


@dataclass
class Box:
    """An axis-aligned body in world pixels. Mutated in place by the movers."""

    x: float
    y: float
    w: float
    h: float


@dataclass
class Pickup:
    """A note or a letter. ``taken`` is cleared on retry rather than reparsed."""

    x: float
    y: float
    w: float
    h: float
    ch: str = ""
    taken: bool = False


@dataclass
class Landing:
    """Which surface stopped a vertical move, if any."""

    ground: bool = False
    ceiling: bool = False
    platform: bool = False


@dataclass
class Marker:
    """Where something starts. Kind is only meaningful for enemies."""

    x: float
    y: float
    kind: str = ""


@dataclass
class Rect:
    x: float
    y: float
    w: float
    h: float


class Tilemap:
    """A parsed level: terrain, plus the things that were sitting in it."""

    def __init__(self, level: dict) -> None:
        self.name: str = str(level["name"])
        self.subtitle: str = str(level.get("subtitle", ""))

        rows = list(level["rows"])
        width = max(len(r) for r in rows)
        self.cols = width
        self.rows = len(rows)
        self.w = self.cols * C.TILE
        self.h = self.rows * C.TILE

        self.spawn = Marker(C.TILE, C.TILE)
        self.exit: Rect | None = None
        self.notes: list[Pickup] = []
        self.letters: list[Pickup] = []
        self.enemies: list[Marker] = []
        self.bells: list[Marker] = []
        self.trolleys: list[Marker] = []

        self.grid: list[list[str]] = []
        for ty, raw in enumerate(rows):
            # Short rows are padded rather than rejected: trimming trailing air
            # from a line should not shift the rest of the level sideways.
            row = list(raw.ljust(width, TILE_AIR))
            for tx in range(width):
                ch = row[tx]
                if ch not in ENTITY_CHARS:
                    continue
                self._place(ch, tx * C.TILE, ty * C.TILE)
                row[tx] = TILE_AIR
            self.grid.append(row)

    def _place(self, ch: str, px: float, py: float) -> None:
        if ch == "S":
            # Bottom-aligned in its tile, so a spawn marker drawn sitting on the
            # ground puts his feet on that ground.
            self.spawn = Marker(px + 1, py + C.TILE - C.PLAYER_H)
        elif ch == "E":
            self.exit = Rect(px, py - C.TILE, C.TILE, C.TILE * 2)
        elif ch == "o":
            self.notes.append(Pickup(px + 4, py + 4, 8, 8))
        elif ch in "TRON":
            self.letters.append(
                Pickup(px + 2, py + 2, C.TILE - 4, C.TILE - 4, ch=ch)
            )
        elif ch == "g":
            self.enemies.append(
                Marker(px, py + C.TILE - C.GARGOYLE_H, "gargoyle")
            )
        elif ch == "f":
            self.enemies.append(Marker(px, py, "flyer"))
        elif ch == "s":
            self.enemies.append(Marker(px, py + C.TILE - C.GARGOYLE_H, "statue"))
        elif ch == "b":
            self.bells.append(
                Marker(px + (C.TILE - C.BELL_W) / 2, py + (C.TILE - C.BELL_H) / 2)
            )
        elif ch == "c":
            self.trolleys.append(Marker(px, py + C.TILE - C.TROLLEY_H))

    # ── Reading the grid ────────────────────────────────────────────

    def tile_at(self, tx: int, ty: int) -> str:
        """The character at a tile.

        Off the left and right edges reads as solid, so a level is walled
        rather than open to an endless sideways fall. Above and below read as
        air: dropping out of the bottom is how a pit kills you.
        """
        if tx < 0 or tx >= self.cols:
            return TILE_SOLID
        if ty < 0 or ty >= self.rows:
            return TILE_AIR
        return self.grid[ty][tx]

    def tile_range(
        self, x: float, y: float, w: float, h: float
    ) -> tuple[int, int, int, int]:
        """Every tile a pixel-space box touches, as (tx0, ty0, tx1, ty1)."""
        return (
            math.floor(x / C.TILE),
            math.floor(y / C.TILE),
            math.floor((x + w - 1) / C.TILE),
            math.floor((y + h - 1) / C.TILE),
        )

    def overlaps(self, x: float, y: float, w: float, h: float, test) -> bool:
        tx0, ty0, tx1, ty1 = self.tile_range(x, y, w, h)
        for ty in range(ty0, ty1 + 1):
            for tx in range(tx0, tx1 + 1):
                if test(self.tile_at(tx, ty)):
                    return True
        return False

    def overlaps_solid(self, x: float, y: float, w: float, h: float) -> bool:
        """Does this box overlap brick? One-way platforms are not solid here."""
        return self.overlaps(x, y, w, h, is_solid_tile)

    def overlaps_water(self, x: float, y: float, w: float, h: float) -> bool:
        return self.overlaps(x, y, w, h, is_water_tile)

    def overlaps_updraft(self, x: float, y: float, w: float, h: float) -> bool:
        """Is any part of this box in rising air? Updraft tiles are not solid."""
        return self.overlaps(x, y, w, h, is_updraft_tile)

    def overlaps_rail(self, x: float, y: float, w: float, h: float) -> bool:
        """Is this box over rail? The trolley only rolls where there is track."""
        return self.overlaps(x, y, w, h, is_rail_tile)

    def has_footing(self, x: float, y: float, w: float, h: float) -> bool:
        """Is there footing directly under this box? Used by ledge-turning."""
        return self.overlaps_solid(x, y + h, w, 1) or self.overlaps(
            x, y + h, w, 1, is_platform_tile
        )

    def updraft_centre(self, x: float, y: float, w: float, h: float) -> float | None:
        """Horizontal centre of the updraft column this box is in, or None."""
        tx0, ty0, tx1, ty1 = self.tile_range(x, y, w, h)
        for ty in range(ty0, ty1 + 1):
            for tx in range(tx0, tx1 + 1):
                if is_updraft_tile(self.tile_at(tx, ty)):
                    return tx * C.TILE + C.TILE / 2
        return None

    def platform_top(
        self, x: float, y: float, w: float, h: float, prev_bottom: float
    ) -> float | None:
        """Top of the highest one-way platform this box has just landed on.

        ``prev_bottom`` is the whole of what makes it one-way: a platform only
        stops a body that was already above it, so you rise through the same
        tile you later stand on.
        """
        tx0, ty0, tx1, ty1 = self.tile_range(x, y, w, h)
        # Scan one row further down than the body occupies. tile_range stops at
        # (y + h - 1), so a body resting EXACTLY on a platform -- bottom flush
        # with the tile top -- never sees the tile holding it up, and sinks: the
        # first frame the tile is out of range, and by the second the playhead
        # has passed it and the came-from-above guard rejects it. Landing from a
        # height hides this, because a fast step crosses the boundary in one
        # move.
        bottom_row = math.floor((y + h) / C.TILE)
        best: float | None = None
        for ty in range(ty0, max(ty1, bottom_row) + 1):
            top = ty * C.TILE
            if prev_bottom > top + 0.001:
                continue  # came from below, or inside
            for tx in range(tx0, tx1 + 1):
                if is_platform_tile(self.tile_at(tx, ty)) and (
                    best is None or top < best
                ):
                    best = top
        return best

    # ── Moving through it ───────────────────────────────────────────

    def move_x(self, box: Box, dx: float) -> bool:
        """Move a body horizontally, stopping flush against brick.

        Stepped under a tile at a time so nothing can tunnel through a wall.
        Normal running never needs it -- 4.2px a frame at most -- but a stomp
        bounce plus knockback can exceed a tile in one step, and passing
        through a wall is the one bug that makes a platformer feel broken
        rather than merely hard.
        """
        if not dx:
            return False
        cap = C.TILE - 1
        step = sign(dx) * min(abs(dx), cap)
        remaining = dx
        hit = False

        while abs(remaining) > 0.0001:
            move = remaining if abs(remaining) < abs(step) else step
            nxt = box.x + move
            if self.overlaps_solid(nxt, box.y, box.w, box.h):
                box.x = (
                    math.floor((nxt + box.w) / C.TILE) * C.TILE - box.w
                    if move > 0
                    else (math.floor(nxt / C.TILE) + 1) * C.TILE
                )
                hit = True
                break
            box.x = nxt
            remaining -= move
        return hit

    def move_y(self, box: Box, dy: float, dropping: bool = False) -> Landing:
        """Move a body vertically. Reports which surface stopped it.

        ``dropping`` suppresses one-way platforms, for a body deliberately
        dropping through one.
        """
        result = Landing()
        if not dy:
            return result

        cap = C.TILE - 1
        step = sign(dy) * min(abs(dy), cap)
        remaining = dy

        while abs(remaining) > 0.0001:
            move = remaining if abs(remaining) < abs(step) else step
            prev_bottom = box.y + box.h
            nxt = box.y + move

            if self.overlaps_solid(box.x, nxt, box.w, box.h):
                if move > 0:
                    box.y = math.floor((nxt + box.h) / C.TILE) * C.TILE - box.h
                    result.ground = True
                else:
                    box.y = (math.floor(nxt / C.TILE) + 1) * C.TILE
                    result.ceiling = True
                break

            if move > 0 and not dropping:
                top = self.platform_top(box.x, nxt, box.w, box.h, prev_bottom)
                if top is not None:
                    box.y = top - box.h
                    result.ground = True
                    result.platform = True
                    break

            box.y = nxt
            remaining -= move
        return result


def overlap(
    ax: float, ay: float, aw: float, ah: float,
    bx: float, by: float, bw: float, bh: float,
) -> bool:
    """Axis-aligned box intersection, as ``entities.js`` defines it."""
    return ax < bx + bw and ax + aw > bx and ay < by + bh and ay + ah > by
